### PR TITLE
Add a time argument that if exists on client is used for FIDO's result table

### LINKED ISSUE #2082: Times on result table on unidown doesn't make sense
At the moment unidown query response shows a table where start and end dates are in increasing day order from the starting date of the query till the number of files(urls) obtained. This works well for instrument that takes one file per day, but it's not correct for cases like imagers where you get multiple images per day.

### PR DESCRIPTION
**Pull request by dpshelio**
This fixes #2082 and it's used in #1807.

The idea is that each client has a `_get_time_for_url()` function so it generates the time ranges for each observation based on the instrument properties. That's the only way I can figure it out to provide the start end date of each url without downloading it.

### CODE REVIEW

**Comment by Cadair** [comment]
Can you show us a comparison of output?

**Comment by dpshelio** [comment]
Sure!
Before
```python
>>> from sunpy.net import Fido
>>> from sunpy.net import attrs as a
>>> results = Fido.search(a.Time('2016/5/18 15:28:00','2016/5/18 16:30:50'), a.Instrument('bbso'), a.Level('fr'))
>>> results
<sunpy.net.fido_factory.UnifiedResponse object at 0x7f6a2641d710>
Results from the BBSOClient:
<class 'sunpy.net.dataretriever.client.QueryResponse'><Table length=3>
     Start Time           End Time              Source        Instrument    Wavelength  
       str19               str19                str21            str4         str15     
------------------- ------------------- --------------------- ---------- ---------------
2016-05-18 00:00:00 2016-05-19 00:00:00 Global Halpha Network       BBSO 6562.8 Angstrom
2016-05-19 00:00:00 2016-05-20 00:00:00 Global Halpha Network       BBSO 6562.8 Angstrom
2016-05-20 00:00:00 2016-05-21 00:00:00 Global Halpha Network       BBSO 6562.8 Angstrom
```
With this:
```python
<sunpy.net.fido_factory.UnifiedResponse object at 0x7f2e659f9f60>
Results from the BBSOClient:
<class 'sunpy.net.dataretriever.client.QueryResponse'><Table length=3>
     Start Time           End Time              Source        Instrument    Wavelength  
       str19               str19                str21            str4         str15     
------------------- ------------------- --------------------- ---------- ---------------
2016-05-18 15:30:25 2016-05-18 15:30:25 Global Halpha Network       BBSO 6562.8 Angstrom
2016-05-18 16:00:33 2016-05-18 16:00:33 Global Halpha Network       BBSO 6562.8 Angstrom
2016-05-18 16:30:46 2016-05-18 16:30:46 Global Halpha Network       BBSO 6562.8 Angstrom
```

**Comment by dpshelio** [comment]
#2084 has also included that function for the available clients

File: sunpy/net/dataretriever/client.py [hunk 1]
```diff
@@ -50,14 +50,14 @@ def __init__(self, map0, url):
         self.phyobs = map0.get('phyobs', "Data not Available")
         self.instrument = map0.get('instrument', "Data not Available")
         self.url = url
-        self.time = TimeRange(map0.get('Time_start'), map0.get('Time_end'))
+        self.time = TimeRange(map0.get('Time_start'), map0.get('Time_end')) if time==None else time
```
Comment (nabobalis) [inline sunpy/net/dataretriever/client.py]: Isn't this a bit ugly?

Comment (Cadair) [inline sunpy/net/dataretriever/client.py]: looks ok to me, what would you propose?

Comment (nabobalis) [inline sunpy/net/dataretriever/client.py]: cant have a full whitespace? also no spaces around the ==?

File: sunpy/net/dataretriever/client.py [hunk 2]
```diff
@@ -248,6 +248,8 @@ def query(self, *args, **kwargs):
         kwergs.update(kwargs)
         urls = self._get_url_for_timerange(
             self.map_.get('TimeRange'), **kwergs)
+        if getattr(self, "_get_time_for_url", None):
+            return QueryResponse.create(self.map_, urls, self._get_time_for_url(urls))
```
Comment (nabobalis) [inline sunpy/net/dataretriever/client.py]: also maybe im just being stupid here, but where is _get_time_for_url defined?

File: sunpy/net/dataretriever/client.py (hunk 1 above)
Comment (Cadair) [inline sunpy/net/dataretriever/client.py]: fair.

File: sunpy/net/dataretriever/client.py (hunk 2 above)
Comment (Cadair) [inline sunpy/net/dataretriever/client.py]: optionally on the client classes.

Comment (nabobalis) [inline sunpy/net/dataretriever/client.py]: Oh, ignore me. I can't read code.

**Pull request merged**